core/src/main/java/io/grpc/SynchronizationContext.java

```diff
@@ -0,0 +1,203 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A synchronization context is a queue of tasks that run in sequence.  It offers following
+ * guarantees:
+ *
+ * <ul>
+ *    <li>Ordering.  Tasks are run in the same order as they are submitted.</li>
+ *    <li>Serialization.  Tasks are run in sequence and establish a happens-before relationship
+ *        between them. </li>
+ *    <li>Non-reentrancy.  If a task running in a synchronization context executes or schedules
+ *        another task in the same synchronization context, the latter task will never run
+ *        inline.  It will instead be queued and run only after the current task has returned.</li>
+ * </ul>
+ *
+ * <p>It doesn't own any thread, and doesn't guarantee which thread the tasks will run on.
+ *
+ * <p>This class is thread-safe.
+ */
+@ThreadSafe
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/4984")
+public class SynchronizationContext implements Executor {
+  private static final Logger log = Logger.getLogger(SynchronizationContext.class.getName());
+
+  private final Object lock = new Object();
+
+  @GuardedBy("lock")
+  private final Queue<Runnable> queue = new ArrayDeque<Runnable>();
+  @GuardedBy("lock")
+  private boolean draining;
+
+  /**
+   * Run all tasks in the queue in the current thread, if no other thread is running this method.
+   * Otherwise do nothing.
+   *
+   * <p>Upon returning, it guarantees that all tasks submitted by {@code #executeLater} before it
+   * have been or will eventually be run, while not requiring any more calls to {@code drain()}.
+   */
+  public final void drain() {
+    boolean drainLeaseAcquired = false;
+    while (true) {
+      Runnable runnable;
+      synchronized (lock) {
+        if (!drainLeaseAcquired) {
+          if (draining) {
+            return;
+          }
+          draining = true;
+          drainLeaseAcquired = true;
+        }
+        runnable = queue.poll();
+        if (runnable == null) {
+          draining = false;
+          break;
+        }
+      }
+      try {
+        runnable.run();
+      } catch (Throwable t) {
+        handleUncaughtThrowable(t);
+      }
+    }
+  }
+
+  /**
+   * Enqueues a task that will be run when {@link #drain} is called.
+   */
+  public final void executeLater(Runnable runnable) {
```

Comment:
It might be good to point out this is useful for adding things from within a lock and then calling drain outside the lock.

Reply:
Done.